### ISSUE TITLE
Routing too eager to avoid traffic lights by using a dangerous U-turn instead

### ISSUE THREAD

**Issue by stragu** [osmandapp/OsmAnd-iOS#928]
I have had this issue several times, where OsmAnd prefers a very akward (and likely prohibited, impossible or dangerous) U-turn rather than go through traffic lights.

In OSM, many bigger intersections double-up the traffic light node, like here: https://www.openstreetmap.org/#map=19/-27.52522/153.18955

This results in OsmAnd trying to make the trip quicker by avoiding them via turning lanes that don't have traffic lights, leading to a U-turn that I assume most users wouldn't prefer.

![image](https://user-images.githubusercontent.com/1747497/99239488-dae20780-2846-11eb-98f6-fe069c2b3806.png)

You can test this example with directions from -27.5284, 153.2837 to -27.4882, 153.0018

Graphhopper and OSRM do not do that when going through the same intersection.

I think this problem comes from a more general issue: **OsmAnd uses U-turns too easily** (for example when missing a turn in the directions, the app often insists that we have to go back with a U-turn). Many would prefer to go around a block rather than do a U-turn. It is often difficult, and extra dangerous to perform a U-turn. And in some areas, it is "prohibited unless signed otherwise", and it is likely that many U-turns recommended by OsmAnd are not legal because of a solid line in the middle of the road, or outright impossible to perform because of a physical barrier that is not included in the database.

OsmAnd iOS version: 3.81

**Comment by vshcherb** [comment]
Yes, this is likely problem with map data. Lack of map data leads to shortcuts which probably don't even exist in reality

**Comment by stragu** [comment]
@vshcherb do you mean that you think there is missing data at that spot?
If that is the case, I disagree. If anything, there is _too much_ data: two traffic lights when crossing the intersection, when in reality one would only stop once. The reality is that this kind of tagging is coming, in which one way has to traffic light nodes but one flow of traffic is only affected by one (and the direction is not tagged yet).

In my report, I mentioned that other routing systems like Graphhopper and OSRM do not use a shortcut at that same location, and go directly through the intersection as expected.

What do you think?

**Comment by vshcherb** [comment]
Other routers simply do not use penalty for traffic_signals which has other issues. We also support traffic_signals groups and there is another issue ( think in Android thread) where we're going to handle it by introducing special penalty for sharp turns

**Comment by stragu** [comment]
Oh that makes a lot more sense, thank you! So I guess you're right, this might be a case of not enough information if the grouping information is missing. But what do you mean by "traffic_signals groups"? I am not aware of a relationship widely used for that.

From the [OSM wiki](https://wiki.openstreetmap.org/wiki/Tag:highway%3Dtraffic_signals), I see this:

> "It is up to the routing software to count nearby signals as one for timing purposes."

Which makes sense, as traffic signals spaced by only a few meters a very likely to work in conjunction. Is that what OsmAnd uses to group traffic signals, or is it an actual tag/relationship?

About the future development around sharp turn penalties: yes please! That would solve quite a few issues with impractical routing, including this one I expect. Looking forward to that!

**Comment by vshcherb** [comment]
OsmAnd supports traffic_signals:direction tag and plan to minimize penalty if traffic signals appear very soon i.e. after 30 m

**Comment by stragu** [comment]
Good to know about the direction tag, and looking forward to the penalty minimisation in future versions. Thanks for the info, @vshcherb :smiley:
